File: sdks/python/kavachos/audit.py
"""Audit log query and export operations."""

from __future__ import annotations

from typing import List, Optional

from kavachos._http import AsyncTransport, SyncTransport
from kavachos.types import AuditEntry, AuditFilters, ExportOptions, PaginatedAuditLogs


class AsyncAuditResource:
    """Async audit log access."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._http = transport

    async def query(
        self,
        filters: Optional[AuditFilters] = None,
    ) -> List[AuditEntry]:
        """Query the audit log.

        Args:
            filters: Optional query filters. Use ``limit`` and ``offset`` for
                pagination.

        Returns:
            List of :class:`~kavachos.types.AuditEntry` objects, newest first.
        """
        params = filters.to_params() if filters else None
        data = await self._http.request("GET", "/audit", params=params)

        # The API may return either a plain list or a paginated wrapper
        if isinstance(data, list):
            return [AuditEntry.from_dict(e) for e in data]

        return PaginatedAuditLogs.from_dict(data).entries

    async def query_paginated(
        self,
        filters: Optional[AuditFilters] = None,
    ) -> PaginatedAuditLogs:
        """Query the audit log and return the full paginated response.

        Args:
            filters: Optional query filters.

        Returns:
            :class:`~kavachos.types.PaginatedAuditLogs` with ``entries`` and
            optional ``total`` count.
        """
        params = filters.to_params() if filters else None
        data = await self._http.request("GET", "/audit", params=params)

        if isinstance(data, list):
            return PaginatedAuditLogs(entries=[AuditEntry.from_dict(e) for e in data])

        return PaginatedAuditLogs.from_dict(data)

    async def export(self, options: Optional[ExportOptions] = None) -> str:
        """Export the audit log as JSON or CSV.

        Args:
            options: Format and date range. Defaults to JSON format.

        Returns:
            Raw export string (JSON text or CSV text depending on format).
        """
        opts = options or ExportOptions()
        params = opts.to_params()
        return await self._http.request_raw("GET", "/audit/export", params=params)


class SyncAuditResource:
    """Sync audit log access."""

    def __init__(self, transport: SyncTransport) -> None:
        self._http = transport

    def query(
        self,
        filters: Optional[AuditFilters] = None,
    ) -> List[AuditEntry]:
        """Query the audit log."""
        params = filters.to_params() if filters else None
        data = self._http.request("GET", "/audit", params=params)

        if isinstance(data, list):
            return [AuditEntry.from_dict(e) for e in data]

        return PaginatedAuditLogs.from_dict(data).entries

    def query_paginated(
        self,
        filters: Optional[AuditFilters] = None,
    ) -> PaginatedAuditLogs:
        """Query the audit log and return the full paginated response."""
        params = filters.to_params() if filters else None
        data = self._http.request("GET", "/audit", params=params)

        if isinstance(data, list):
            return PaginatedAuditLogs(entries=[AuditEntry.from_dict(e) for e in data])

        return PaginatedAuditLogs.from_dict(data)

    def export(self, options: Optional[ExportOptions] = None) -> str:
        """Export the audit log as JSON or CSV."""
        opts = options or ExportOptions()
        params = opts.to_params()
        return self._http.request_raw("GET", "/audit/export", params=params)

File: sdks/go/audit.go
package kavachos

import (
	"context"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

// AuditResource provides access to the audit log.
type AuditResource struct {
	tp *transport
}

// Query retrieves audit log entries. Pass nil to return all entries.
func (r *AuditResource) Query(ctx context.Context, filters *AuditFilters) ([]AuditEntry, error) {
	result, err := r.QueryPaginated(ctx, filters)
	if err != nil {
		return nil, err
	}
	return result.Entries, nil
}

// QueryPaginated retrieves audit log entries and includes the total count.
func (r *AuditResource) QueryPaginated(ctx context.Context, filters *AuditFilters) (*PaginatedAuditLogs, error) {
	params := auditFiltersToParams(filters)
	var raw json.RawMessage
	if err := r.tp.doJSON(ctx, "GET", "/audit", nil, params, &raw); err != nil {
		return nil, err
	}
	data := unwrapAPIData(raw)

	// The API may return either a plain list or a { entries, total } object.
	if len(data) > 0 && data[0] == '[' {
		var entries []AuditEntry
		if err := json.Unmarshal(data, &entries); err != nil {
			return nil, fmt.Errorf("kavachos: decode audit entries: %w", err)
		}
		return &PaginatedAuditLogs{Entries: entries}, nil
	}

	var paginated PaginatedAuditLogs
	if err := json.Unmarshal(data, &paginated); err != nil {
		return nil, fmt.Errorf("kavachos: decode paginated audit: %w", err)
	}
	return &paginated, nil
}

// Export exports audit log data as a raw string.
// The format is controlled by ExportOptions.Format (JSON or CSV).
// Pass nil to use the default format (JSON).
func (r *AuditResource) Export(ctx context.Context, opts *ExportOptions) (string, error) {
	params := exportOptionsToParams(opts)
	return r.tp.doRaw(ctx, "GET", "/audit/export", params)
}

func auditFiltersToParams(f *AuditFilters) map[string]string {
	if f == nil {
		return nil
	}
	params := make(map[string]string)
	if f.AgentID != nil {
		params["agentId"] = *f.AgentID
	}
	if f.UserID != nil {
		params["userId"] = *f.UserID
	}
	if f.Since != nil {
		params["since"] = *f.Since
	}
	if f.Until != nil {
		params["until"] = *f.Until
	}
	if len(f.Actions) > 0 {
		params["actions"] = strings.Join(f.Actions, ",")
	}
	if f.Result != nil {
		params["result"] = string(*f.Result)
	}
	if f.Limit != nil {
		params["limit"] = strconv.Itoa(*f.Limit)
	}
	if f.Offset != nil {
		params["offset"] = strconv.Itoa(*f.Offset)
	}
	if len(params) == 0 {
		return nil
	}
	return params
}

func exportOptionsToParams(opts *ExportOptions) map[string]string {
	params := make(map[string]string)
	if opts == nil {
		params["format"] = string(ExportFormatJSON)
		return params
	}
	if opts.Format == "" {
		params["format"] = string(ExportFormatJSON)
	} else {
		params["format"] = string(opts.Format)
	}
	if opts.Since != nil {
		params["since"] = *opts.Since
	}
	if opts.Until != nil {
		params["until"] = *opts.Until
	}
	return params
}
